pkg/koordlet/statesinformer/states_device.go

```diff
@@ -0,0 +1,224 @@
+		return nil
+	}
+	if len(nodeResource.Metric.GPUs) == 0 {
+		klog.V(5).Info("no gpu device found")
+		return nil
+	}
+	var deviceInfos []schedulingv1alpha1.DeviceInfo
+	for _, gpu := range nodeResource.Metric.GPUs {
+		health := true
+		s.gpuMutex.RLock()
+		if _, ok := s.unhealthyGPU[gpu.DeviceUUID]; ok {
+			health = false
+		}
+		s.gpuMutex.RUnlock()
+		deviceInfos = append(deviceInfos, schedulingv1alpha1.DeviceInfo{
+			UUID:   gpu.DeviceUUID,
+			Minor:  gpu.Minor,
+			Type:   schedulingv1alpha1.GPU,
+			Health: health,
+			Resources: map[corev1.ResourceName]resource.Quantity{
+				extension.GPUCore:        *resource.NewQuantity(100, resource.DecimalSI),
+				extension.GPUMemory:      gpu.MemoryTotal,
+				extension.GPUMemoryRatio: *resource.NewQuantity(100, resource.DecimalSI),
+			},
+		})
+	}
+	return deviceInfos
+}
+
+func (s *statesInformer) initGPU() bool {
+	if ret := nvml.Init(); ret != nvml.SUCCESS {
+		if ret == nvml.ERROR_LIBRARY_NOT_FOUND {
+			klog.Warning("nvml init failed, library not found")
+			return false
+		}
+		klog.Warningf("nvml init failed, return %s", nvml.ErrorString(ret))
+		return false
+	}
+	return true
+}
+
+func (s *statesInformer) gpuHealCheck(stopCh <-chan struct{}) {
+	count, ret := nvml.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		klog.Errorf("unable to get device count: %v", nvml.ErrorString(ret))
+		return
+	}
+	if count == 0 {
+		klog.Errorf("no gpu device found")
+		return
+	}
+	devices := []string{}
+	for deviceIndex := 0; deviceIndex < count; deviceIndex++ {
+		gpudevice, ret := nvml.DeviceGetHandleByIndex(deviceIndex)
+		if ret != nvml.SUCCESS {
+			klog.Errorf("unable to get device at index %d: %v", deviceIndex, nvml.ErrorString(ret))
+			continue
+		}
+		uuid, ret := gpudevice.GetUUID()
+		if ret != nvml.SUCCESS {
+			klog.Errorf("failed to get device uuid at index %d, err: %v", deviceIndex, nvml.ErrorString(ret))
+		}
+		devices = append(devices, uuid)
+	}
+	unhealthyChan := make(chan string)
+	go checkHealth(stopCh, devices, unhealthyChan)
+	klog.Info("start to do gpu health check")
+	for d := range unhealthyChan {
+		// FIXME: there is no way to recover from the Unhealthy state.
+		s.gpuMutex.Lock()
+		s.unhealthyGPU[d] = struct{}{}
+		s.gpuMutex.Unlock()
+		klog.Infof("get a unhealthy gpu %s", d)
+	}
+}
+
+// check status of gpus, and send unhealthy devices to the unhealthyDeviceChan channel
+func checkHealth(stopCh <-chan struct{}, devs []string, xids chan<- string) {
+	eventSet, ret := nvml.EventSetCreate()
+	if ret != nvml.SUCCESS {
+		klog.Errorf("failed to create event set, err: %v", nvml.ErrorString(ret))
+		os.Exit(1)
+	}
+	defer eventSet.Free()
+
+	for _, d := range devs {
+		device, ret := nvml.DeviceGetHandleByUUID(d)
+		if ret != nvml.SUCCESS {
+			klog.Errorf("failed to get device %s, err: %v", d, nvml.ErrorString(ret))
+			continue
+		}
+		ret = nvml.DeviceRegisterEvents(device, nvml.EventTypeXidCriticalError, eventSet)
```

Comment:
the whole health logic is little bit different from: https://github.com/NVIDIA/k8s-device-plugin/blob/master/internal/rm/health.go,   please make sure the diff is necessory.

Reply:
+1

Reply:
updated